Gender to Flow

seldridge opened the pull request:
In light of the discussion around #943, this implements a structural rename of Working IR's `Gender` to `Flow`. I'm attempting to maintain backwards compatibility by deprecating anything gender-related that's public and providing implicit conversions where appropriate.

In summary, the following are the new names, which are assumedly more descriptive on what is going on. Nonetheless, I'm happy to bikeshed:
  - `gender        -> flow`
  - `MALE          -> SourceFlow`
  - `FEMALE        -> SinkFlow`
  - `BIGENDER      -> DuplexFlow`
  - `UNKNOWNGENDER -> UnknownFlow`

Fixes #943.

seldridge commented:
ping @azidar

jackkoenig commented on src/main/scala/firrtl/passes/Checks.scala:
```diff
@@ -582,66 +582,83 @@ object CheckGenders extends Pass {
+    case BIGENDER => "sourceOrSink"
+  }
+
+  def run(c: Circuit): Circuit = CheckGenders.run(c)
```
This looks like an infinite recursion to me

seldridge replied:
I prefer to call this "Big Stick Deprecation Diplomacy."

...


(Fixed in rebase.)

jackkoenig replied:
😂

seldridge commented:
Updated deprecation messages and added deprecated `ResolveGenders` methods. A cursory glance seems that adding a deprecated copy method is complicated. I'm punting on that here, but can dig into it if you think otherwise.

jackkoenig commented:
I don't think there's anything we can do about `copy`, so I think we're good to go!
